src/documents/models.py

```diff
@@ -267,17 +277,90 @@ def __str__(self):
         return str(created)
 
     @property
-    def source_path(self):
+    def source_filename(self):
+        if self.filename is None:
+            self.filename = self.generate_source_filename()
+
+        return self.filename
+
+    @staticmethod
+    def many_to_dictionary(field):
+        # Converts ManyToManyField to dictionary by assuming, that field
+        # entries contain an _ or - which will be used as a delimeter
+        mydictionary = dict()
+
+        for t in field.all():
+            # Find delimeter
+            delimeter = t.name.find('_')
+
+            if delimeter is -1:
+                delimeter = t.name.find('-')
+
+            if delimeter is -1:
+                continue
+
+            key = t.name[:delimeter]
+            value = t.name[delimeter+1:]
+
+            mydictionary[slugify(key)] = slugify(value)
+
+        return mydictionary
+
+    @staticmethod
+    def many_to_list(field):
+        # Converts ManyToManyField to list
+        mylist = list()
+
+        for t in field.all():
+            mylist.append(slugify(t))
+
+        return mylist
 
-        file_name = "{:07}.{}".format(self.pk, self.file_type)
+    def generate_source_filename(self):
+        # Create filename based on configured format
+        if settings.PAPERLESS_FILENAME_FORMAT is not None:
+            path = settings.PAPERLESS_FILENAME_FORMAT.format(
+                   correspondent=slugify(self.correspondent),
+                   title=slugify(self.title),
+                   created=slugify(self.created),
+                   added=slugify(self.added),
+                   tag=defaultdict(str, self.many_to_dictionary(self.tags)),
+                   tags=self.many_to_list(self.tags))
```

Comment:
Index-out-of-bounds errors are not handled here, right? This means that if a scanned document does not have any tags applied (or simply not as many as the user references in their format-string) this will throw an exception, breaking document consumption.

This should probably be handled differently, although I'm not yet sure what good behaviour would be here.

Reply:
I have added handling - admittedly in a rather clumsy way but workable and tested through unit tests. What do you think?